mongo/integration/unified/logger_verification.go

```diff
@@ -0,0 +1,274 @@
+
+	// Top level data does not have to be 1-1 with the expectation, there
+	// are a number of unrequired fields that may not be present on the
+	// expected document.
+	if err := verifyValuesMatch(ctx, rawExp, rawAct, true); err != nil {
+		return fmt.Errorf("%w: %v", errLogDocumentMismatch, err)
+	}
+
+	return nil
+}
+
+// clientLogMessages is a struct representing the expected "LogMessages" for a
+// client.
+type clientLogMessages struct {
+	Client      string        `bson:"client"`
+	LogMessages []*logMessage `bson:"messages"`
+}
+
+// validateClientLogMessages will validate a single "clientLogMessages" object
+// and return an error if it is invalid, i.e. not testable.
+func validateClientLogMessages(ctx context.Context, log *clientLogMessages) error {
+	if log.Client == "" {
+		return fmt.Errorf("client is required")
+	}
+
+	if len(log.LogMessages) == 0 {
+		return fmt.Errorf("log messages are required")
+	}
+
+	for _, message := range log.LogMessages {
+		if err := validateLogMessage(ctx, message); err != nil {
+			return fmt.Errorf("message is invalid: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// validateExpectLogMessages will validate a slice of "clientLogMessages"
+// objects and return the first error encountered.
+func validateExpectLogMessages(ctx context.Context, logs []*clientLogMessages) error {
+	seenClientNames := make(map[string]struct{}) // Check for client duplication
+
+	for _, log := range logs {
+		if err := validateClientLogMessages(ctx, log); err != nil {
+			return fmt.Errorf("client is invalid: %v", err)
+		}
+
+		if _, ok := seenClientNames[log.Client]; ok {
+			return fmt.Errorf("duplicate client: %v", log.Client)
+		}
+
+		seenClientNames[log.Client] = struct{}{}
+	}
+
+	return nil
+}
+
+// logMessageValidator defines the expectation for log messages across all
+// clients.
+type logMessageValidator struct {
+	testCase *TestCase
+	err      chan error
+}
+
+// newLogMessageValidator will create a new "logMessageValidator" from a test
+// case.
+func newLogMessageValidator(testCase *TestCase) (*logMessageValidator, error) {
+	if testCase == nil {
+		return nil, fmt.Errorf("test case is required")
+	}
+
+	if testCase.entities == nil {
+		return nil, fmt.Errorf("entities are required")
+	}
+
+	validator := &logMessageValidator{
+		testCase: testCase,
+		err:      make(chan error, len(testCase.entities.clients())),
+	}
+
+	return validator, nil
+}
+
+type actualLogQueues map[string]chan orderedLogMessage
+
+func (validator *logMessageValidator) expected(ctx context.Context) ([]*clientLogMessages, actualLogQueues) {
+	clients := entities(ctx).clients()
+
+	expected := make([]*clientLogMessages, 0, len(validator.testCase.ExpectLogMessages))
+	actual := make(actualLogQueues, len(clients))
+
+	for _, clientLogMessages := range validator.testCase.ExpectLogMessages {
+		clientName := clientLogMessages.Client
+
+		clientEntity, ok := clients[clientName]
+		if !ok {
+			continue // If there is no entity for the client, skip it.
```

Comment:
Would this not be an error?

Reply:
I'm not sure, I would argue that if there is no client then there is nothing to verify one way or another. No operations would occur.